core/src/main/java/org/infinispan/transaction/impl/WriteSkewHelper.java

```diff
@@ -57,52 +61,62 @@ public static EntryVersionsMap performWriteSkewCheckAndReturnNewVersions(Version
                }
                VersionedRepeatableReadEntry entry = (VersionedRepeatableReadEntry) cacheEntry;
 
-               if (entry.performWriteSkewCheck(dataContainer, segment, persistenceManager, context,
-                                               prepareCommand.getVersionsSeen().get(k), versionGenerator, timeService)) {
-                  IncrementableEntryVersion oldVersion = (IncrementableEntryVersion) entry.getMetadata().version();
-                  IncrementableEntryVersion newVersion = entry.isCreated() || oldVersion == null
-                        ? versionGenerator.generateNew()
-                        : versionGenerator.increment(oldVersion);
-                  uv.put(k, newVersion);
-               } else {
-                  // Write skew check detected!
-                  throw new WriteSkewException("Write skew detected on key " + k + " for transaction " +
-                                                     context.getCacheTransaction(), k);
-               }
+               CompletionStage<Boolean> skewStage = entry.performWriteSkewCheck(entryLoader, segment, context,
+                     prepareCommand.getVersionsSeen().get(k), versionGenerator);
+               aggregateCompletionStage.dependsOn(skewStage.thenAccept(passSkew -> {
```

Comment:
Could we not complete the `performWriteSkewCheck` CompletionStage exceptionally if the check fails? That way we don't duplicate the exception throwing logic twice in this class, we just handle a success.

Reply:
Just to confirm what you are saying, you mean to change `VersionedRepeatableReadEntry#performWriteSkewCheck` to propagate the exception in the `CompletionStage<Void>` return value?

Reply:
But thinking about it, with the current method names that doesn't really make sense. Something like `performIfNoWriteSkew` would better fit the semantics of what I'm suggesting, but I think I'm just splitting hairs now. So happy to stick with your original design.

Reply:
Yeah tbh it was like this before. I wanted to consolidate, but didn't want to mix up stuff with the map being written to or not.